libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,427 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+
+        if( bytesSent == MQTT_PUBLISH_ACK_PACKET_SIZE )
+        {
+            pContext->controlPacketSent = true;
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            packetType,
+                                            MQTT_SEND );
+
+            if( newState != expectedNextState )
+            {
+                status = MQTTIllegalState;
+            }
+        }
+        else
+        {
+            status = MQTTSendFailed;
+        }
+    }
+
+    if( ( status == MQTTSuccess ) && ( newState != MQTTStateNull ) )
+    {
+        *pPublishState = newState;
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send a keep alive PINGREQ if the keep alive interval has elapsed.
+ *
+ * @param[in] pContext Initialized MQTT Context.
+ *
+ * @return #MQTTKeepAliveTimeout if a PINGRESP is not received in time,
+ * #MQTTSendFailed if the PINGREQ cannot be sent, or #MQTTSuccess.
+ */
+static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    assert( pContext != NULL );
+    MQTTGetCurrentTimeFunc_t getCurrentTime = pContext->callbacks.getTime;
+    uint32_t now = getCurrentTime();
+    uint32_t keepAliveMs = 1000U * ( uint32_t ) pContext->keepAliveIntervalSec;
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( keepAliveMs && ( ( now - pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( ( now - pContext->pingReqSendTimeMs ) > pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+            /* Sending the packet may have taken some time. Recalculate. */
+            pContext->pingReqSendTimeMs = getCurrentTime();
+            pContext->waitingForPingResp = true;
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Handle received MQTT PUBLISH packet.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] pIncomingPacket Incoming packet.
+ *
+ * @return MQTTSuccess, MQTTIllegalState or deserialization error.
+ */
+static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
+                                           MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    MQTTPublishInfo_t publishInfo;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    status = MQTT_DeserializePublish( pIncomingPacket, &packetIdentifier, &publishInfo );
+    IotLogInfoWithArgs( "Publish packet deserialized with result %d.", status );
```

Comment:
`De-serialized incoming PUBLISH packet: DeserializerResult=%d`

Reply:
Also, please log the stringified version of the result. The MQTT library needs a `strerror` utility

Reply:
Changed the log. Agree with strerror but think it's outside the scope of this PR.